hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/savepoint/SavepointActionExecutor.java

```diff
@@ -145,10 +132,41 @@ public HoodieSavepointMetadata execute() {
           .saveAsComplete(
               true, instantGenerator.createNewInstant(HoodieInstant.State.INFLIGHT, HoodieTimeline.SAVEPOINT_ACTION, instantTime), Option.of(metadata),
               savepointCompletedInstant -> table.getMetaClient().getTableFormat().savepoint(savepointCompletedInstant, table.getContext(), table.getMetaClient(), table.getViewManager()));
-      LOG.info("Savepoint " + instantTime + " created");
+      LOG.info("Savepoint {} created", instantTime);
       return metadata;
     } catch (HoodieIOException e) {
       throw new HoodieSavepointException("Failed to savepoint " + instantTime, e);
     }
   }
+
+  @VisibleForTesting
+  String getLastCommitRetained() {
+    Option<HoodieInstant> cleanInstant = table.getCleanTimeline().lastInstant();
+    // if there are no clean instants, we can use the first completed commit
+    if (cleanInstant.isEmpty()) {
+      return table.getCompletedCommitsTimeline().firstInstant().get().requestedTime();
+    }
+    return cleanInstant.map(instant -> {
+      try {
+        if (instant.isCompleted()) {
+          return table.getActiveTimeline().readCleanMetadata(instant)
+              .getEarliestCommitToRetain();
+        } else {
+          // clean is pending or inflight
+          return table.getActiveTimeline().readCleanerPlan(
+                  instantGenerator.createNewInstant(REQUESTED, instant.getAction(), instant.requestedTime()))
+              .getEarliestInstantToRetain().getTimestamp();
+        }
+      } catch (IOException e) {
+        throw new HoodieSavepointException("Failed to savepoint " + instantTime, e);
+      }
+    }).orElseGet(() ->
+      // If there is no earliest commit to retain, but there are clean instants on the timeline, we must assume KEEP_LATEST_FILE_VERSIONS policy is used.
+      // In that case, the last commit before the clean instant is the last commit guaranteed to be retained.
+      table.getActiveTimeline().getWriteTimeline().filterCompletedInstants()
+          .filter(instant -> compareTimestamps(instant.requestedTime(), LESSER_THAN_OR_EQUALS, cleanInstant.get().requestedTime()))
```

Comment:
This is not consistent with the comment above "Cannot allow savepoint time on a commit that could have been cleaned", should it be `GREATER_THAN` instead?

Reply:
This is searching for the last commit before the clean because that is the last known valid point in time where all data files are guaranteed to be present since the clean will not touch the files in the current view of the table when it runs its plan.

Reply:
yeah, the `GREATER_THAN` check is still in line82 so we should be good, actually here you want to find the smallest instant that is greater than the latest clean instant. You can reuse the `getCompletedCommitsTimeline()` in line 147 and get the last instant of it, if is greater than then last clean instant, return it otherwise just reuturn the clean instant.

```java
HoodieInstant lastCommit = getCompletedCommitsTimeline().lastInstant().orElse(null);
return lastCommit != null && lastCommit.greaterThan(cleanInstant)? lastCommit  : cleanInstant;
```

Reply:
I am looking for the last commit before the latest clean instant, not the one after it. Is there any good utility for that?

Reply:
Updated to use the clean instant instead of throwing an exception